Fix clear year and month field when converting to biblatex

catarinagomespt opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->
Fixes https://github.com/JabRef/jabref/issues/6224

After Cleanup entries with the selected option -> "Convert to biblatex format (for example, move the value of the 'journal' field to 'journal tile')
Should have two behaviours:
- if the field date is filled: remains the date and removes the year and month fields;
- if the field date is blank: sets the date value and removes the year and month fields.


<!-- 
- Go through the list below. If a task has been completed, mark it done by using `[x]`.
- Please don't remove any items, just leave them unchecked if they are not applicable.
-->

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

catarinagomespt commented:
> Looks good to me. Thank you very much.
> Please add a changelog entry and this PR should be mergeable.
@calixtus 
Done. I changed and pushed Changelog.md file with the description of the issue solved.
Please confirm. Thanks

tobiasdiez commented on src/main/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanup.java:
```diff
@@ -39,6 +39,12 @@
+        } else {
+            // If field date is filled it should be removed the year and month fields
+            entry.getFieldOrAlias(StandardField.DATE).ifPresent(newDate -> {
+                entry.clearField(StandardField.YEAR).ifPresent(changes::add);
```
I think there should also be a check that the information in the date field coincides with the one in year & month. For example, if `date = {2004-05}` but `year = {2005}` the year field shouldn't be deleted.

calixtus replied:
Is this really the job of this class? The code assures, that there is something in the date field. I don't think that the cleanup class has to check, if the conversion was done right...

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -54,6 +54,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where JabRef switched to discrete graphics under macOS [#5935](https://github.com/JabRef/jabref/issues/5935)
 - We fixed an issue where the Preferences entry preview will be unexpected modified leads to Value too long exception [#6198](https://github.com/JabRef/jabref/issues/6198)
 - We fixed an issue where custom jstyles for Open/LibreOffice would only be valid if a layout line for the entry type `default` was at the end of the layout section [#6303](https://github.com/JabRef/jabref/issues/6303)
+- We fixed an issue where clear year and month field when converting to biblatex [#6224](https://github.com/JabRef/jabref/issues/6224)
```
maybe "where year and month fields were not cleared" ?

catarinagomespt replied:
Done.

silverhorse commented:
ConvertToBiblatexCleanupTest.cleanupWithDateAlreadyPresentDoesNothing() needs to be fixed to assert that year and month are empty when all 3 date, year, month are present. May be add more test as per changes requested. I started fixing the issue at the same time as you @catarinagomespt  :)

catarinagomespt commented:
@tobiasdiez 
**Summarize:**
1. Added additional check before cleaning month/year field:
- Check if the date field is filled 
- Check if the date field is equal to year
- Check if the year field is not empty

2. Added a unit test for the change. 
Thanks!

catarinagomespt commented:
Thanks @silverhorse. I already saw the ConvertToBiblatexCleanupTest tests.

calixtus commented on src/main/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanup.java:
```diff
@@ -39,8 +39,18 @@
+        } else {
+            // If date field is filled and is equal to year it should be removed the year and month fields
+            entry.getFieldOrAlias(StandardField.DATE).ifPresent(date -> {
+                entry.getFieldOrAlias(StandardField.YEAR).ifPresent(yearl -> {
```
Can you give the variable `year1` a more meaningful name? Sure, everybody understands that, but nevertheless, we try to write exemplary code in JabRef (I know, it's still a loooooong way to go).

catarinagomespt replied:
you're right, I forgot to change

catarinagomespt replied:
Done.

catarinagomespt replied:
Done.

tobiasdiez commented on src/test/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanupTest.java:
```diff
@@ -46,6 +46,20 @@ public void cleanupWithDateAlreadyPresentDoesNothing() {
+
+        assertEquals(Optional.empty(), entry.getField(StandardField.YEAR));
+        assertEquals(Optional.empty(), entry.getField(StandardField.MONTH));
+        assertEquals(Optional.of("2011"), entry.getField(StandardField.DATE));
```
This should be `2011-01` to include the month, right?

tobiasdiez commented on src/main/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanup.java:
```diff
@@ -39,8 +39,18 @@
+            entry.getFieldOrAlias(StandardField.DATE).ifPresent(date -> {
+                entry.getFieldOrAlias(StandardField.YEAR).ifPresent(checkYear -> {
+                    // Get the year of the date field (4 digits).
+                    if (date.substring(0, 4).equals(checkYear)) {
```
Adding to @koppor's comment, we have a date class which you can use to parse the date in order to get the year AND the month.

catarinagomespt replied:
@koppor @tobiasdiez 
Please confirm that I understood correctly:
**Acceptance Criteria  1:**
- If the month + year fields are equal to the month + year in the date field, must clear the month and the year.
**Start:** 
date = "2020-05"
month = "05"
year = "2020"
**Result:** 
date = "2020-05"
month = ""
year = ""

**Acceptance Criteria 2:**
- If the month field is different from the month in the date field, does nothing
**Start:** 
date = "2020-05"
month = "04"
year = "2020"
**Result:** 
date = "2020-05"
month = "04"
year = "2020"

**Acceptance Criteria 3:**
- If the year field is different from the year in the date field, does nothing
**Start:** 
date = "2020-05"
month = "05"
year = "2019"
**Result:** 
date = "2020-05"
month = "05"
year = "2019"

**Acceptance Criteria 4:**
- If the date field has a blank/empty value, must set the date field with year and month values and clear the month + year fields.
**Start:** 
date = ""
month = "05"
year = "2020"
**Result:** 
date = "2020-05"
month = ""
year = ""

Thanks

catarinagomespt replied:
@koppor @tobiasdiez 
**Another question:**
What should happen if we have the following situation?
date = "2020"
month = "04"
year = "2020"
**Result:** ???

tobiasdiez replied:
I think in scenario 2 + 3 the correct information (the month in 2, the year in 3) can be deleted leaving only differing information behind.

catarinagomespt replied:
@tobiasdiez 
Just to make sure I got it right.
**Start:**
date = "2020-05"
month = "04"
year = "2020"

**Result:**
date = "2020-05"
month = "04"
year = ""

Please, confirms that the result is correct.

Siedlerchr replied:
For inconsistent data: Don't overwrite/move. Keep as is

catarinagomespt replied:
@tobiasdiez 
> Adding to @koppor's comment, we have a date class which you can use to parse the date in order to get the year AND the month.

What is the class name?

tobiasdiez replied:
https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/model/entry/Date.java

koppor commented:
Tests currently failing. Think @catarinagomespt is working on it.

![grafik](https://user-images.githubusercontent.com/1366654/81640295-7325ac80-941e-11ea-85a9-430fb7d1f7cb.png)

tobiasdiez commented:
@catarinagomespt why did you removed the new tests? I think they were pretty good and important to verify that the implementation is also working in certain edge cases.

Siedlerchr commented on src/main/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanup.java:
```diff
@@ -39,8 +41,26 @@
+        } else {
+            // If the year from date field is filled and equal to year it should be removed the year field
+            entry.getFieldOrAlias(StandardField.DATE).ifPresent(date -> {
+                entry.getFieldOrAlias(StandardField.YEAR).ifPresent(year -> {
```
One of the Date.parse method overloads accepts already Optional as parameter, so you can make your code a bit easier and remove the ifPressent checks for Year and month

so you would have` checkDate = Date.parse(entry.getFieldOrAlias(StandardField.YEAR),  entry.getFieldOrAlias(StandardField.Month), Optional.empty())`

catarinagomespt replied:
@Siedlerchr Thanks for the tip. Already did that.

Siedlerchr commented on src/main/java/org/jabref/logic/cleanup/ConvertToBiblatexCleanup.java:
```diff
@@ -39,8 +41,19 @@
+                Optional<Date> newDate = Date.parse(date);
+                Optional<Date> checkDate = Date.parse(entry.getFieldOrAlias(StandardField.YEAR),
+                        entry.getFieldOrAlias(StandardField.MONTH), Optional.empty());
+
```
Just a quick note,
The Optional equals also returns true if both are empty.
That could be the case when there is a value in all fields, but it's invalid., eg maybe some random letters.
Could you please add a test case for that as well. Then you should have all cases covered. I think in practice this doesn't matter but code wise it would be nice to have this edge case covered as well. 

Otherwise it's looking good to me! Thanks again for your ongoing contribution 

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#equals-java.lang.Object-

catarinagomespt replied:
@Siedlerchr You are right, if both fields are empty returns true. I added a new test case to cover it.

catarinagomespt commented on src/main/java/org/jabref/model/entry/Date.java:
```diff
@@ -54,7 +54,8 @@ public Date(TemporalAccessor date) {
                 "d.M.uuuu",
-                "uuuu.M.d", "uuuu");
+                "uuuu.M.d", "uuuu",
+                "MMM, uuuu");
```
Done. Please confirm.
